File: Problems/0124. (H) Binary Tree Maximum Path Sum/Python3/solution.py
class Solution:
  def maxPathSum(self, root: Optional[TreeNode]) -> int:
    # recursive DFS

    # overall max path sum
    self.max_sum = float('-inf')

    def recurse(node):
      # recursively get left and right path sums
      left_sum = recurse(node.left) if node.left else 0
      right_sum = recurse(node.right) if node.right else 0
      # see whether path sum in subtree rooted at node is larger than previous max
      self.max_sum = max(self.max_sum, left_sum+node.val+right_sum)
      ret = max(node.val+left_sum, node.val+right_sum)
      return ret if ret > 0 else 0

    recurse(root)

    return self.max_sum

File: Problems/0124. (H) Binary Tree Maximum Path Sum/Java/solution.java
class Solution {
  // overall max path sum
  int max_sum = Integer.MIN_VALUE;

  public int recurse(TreeNode node) {
    // recursively get left and right path sums
    int left_sum = 0;
    int right_sum = 0;
    if (node.left != null)
      left_sum = recurse(node.left);
    if(node.right != null)
      right_sum = recurse(node.right);
    // see whether path sum in subtree rooted at node is larger than previous max
    this.max_sum = Math.max(this.max_sum, left_sum+node.val+right_sum);
    int ret = Math.max(left_sum+node.val, node.val+right_sum);
    return ret > 0 ? ret : 0;
  }


  public int maxPathSum(TreeNode root) {
    recurse(root);

    return this.max_sum;
  }
}
